IPC >> Shells

jack-fractal opened the pull request:
So here's this thing.

The thread for what it's trying to do is here: http://aurorastation.org/forums/viewtopic.php?f=18&t=2627

This touches a lot of stuff, and it's a lot of stuff that nobody much wants to deal with because it's awful. The two big ones are update_body, the function that creates the characters base sprite before their clothing and items are layered on top, and the update_preview_icon function, which creates the preview icon for use in the character creation window. 

Both these procedures were enormous messes. They're still a mess, but now at they're a different mess. Preview icon creation has been delegated to a series of datum classes for the clothing layer, and to the species object for mob icon construction. Organ datums can now delegate their own icon retrieval to their 'covering' object, which is what allows the synthetic skin and paint system to work.

I also had to modify, a little, how surgery and removing limbs works. I needed to do this in order to allow for posibrains/mmi's to be removed from machine-people's skulls. This was really awkward to make work. 

There are a bunch of other little things. I had to change the save file to allow you to save your brain type, and change the system for creating cyborg limbs. I added an icon file for the synthetic skin items, created a greyscale copy of the human file, and edited all of the IPC hairstyles to remove their big computer head.

Despite being firmly in the camp that thinks that IPC's badly need a nerf in terms of damage reduction, I haven't touched the actual mechanics of the race at all. These are still the same IPC's, they're just in ~~pog form~~ different colors.

One gameplay related thing I did modify was robot construction. Robots no longer use flashes to create robot eyes. You use a robot camera component. You can also make new shells from robot bodies. The construction process is identical except that for a cyborg you add a new item called a 'law computer' to the cyborg's head before attaching it. This makes shell construction an expensive, but _fast_, alternative to cloning.

This is a big sprawling changelist. ZipZero and I spent a good eight hours on Wednesday doing bug-testing but there will probably still be bugs.

skull132 commented:
Question: how does this affect the manufacturing of normal cyborgs, beyond the requirement to use a robot camera component? I noticed a law computer being a thing, would this spawn along with the chassis (as do all normal components), or does it need to be specially inserted? I ask because there is no mention of this component in your logged changes, and yet it exists, code wise.

jack-fractal commented:
Ah, yes. I should have clarified that. The law computer is a stand-in right now. It determines whether the exoskeleton becomes a normal cyborg or a shell when you insert an mmi or posibrain.

The head has a law computer in it? You create a normal cyborg.
No law computer? You make a shell. 

As a next step, I plan on making law computers attach directly to MMI's, so that the chassis you're in doesn't determine if you have laws or not. I didn't include that in this changelist because to make that viable I'd need to also revise the robot access system, which is a big complicated change all on it's own.

skull132 commented:
As an update, I read over the code and gained an understanding for most of the changes (a lot of it was swapping one mess for another, as you said :/). They're good. It's just a matter of testing it now, which I can probably do tomorrow evening, or @SoundScopes can do while I'm occupied. And we should be on track for pulling this around Tuesday, should no issues creep up.

jack-fractal commented:
Thanks for the update. I would be surprised honestly if there _were_ no bugs in this. We tested as much as we could, ZipZero and I, but I expect there could be issues in edge cases that we missed. When bugs show up, I'll fix them.

The stuff about human limb icons is pretty bad, but I'm pretty happy with how the clothing preview stuff worked out.

skull132 commented:
I doubt we will find any more, as we're going to be on the look out for ones that stop immediate merging and affect more massive things (items which I'd expect you to have already weeded out, but we'll see regardless). But when we put code live, I have a feeling the community will find some more. They're quite good at this.

skull132 commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -112,132 +207,73 @@
+	
+/obj/item/robot_parts/robot_suit/proc/create_shell(obj/item/device/mmi/brain as obj)
+	var/mob/living/carbon/human/machine/new_shell = new(src.loc)
+	var/key=brain.brainmob.mind.key
```
What is this? It reports as being unused. Can delete/comment out, or does it actually serve a purpose?

jack-fractal replied:
Ah, yeah. That's a piece of development leftovers. Feel free to delete it.

skull132 commented:
Okay, found two things. I was fiddling with the character creation settings, giving them a once-over. Ran into two issues:
* First off, the "Brain" on my save is empty. Regardless of what I try to do.
* Secondly, the "Scales" option doesn't seem to be producing desirable results: http://puu.sh/iRGA9/21bb07c030.png _in the preview window_. It loads into the game just fine, though, as an Unathi. Also with the scales setting, you have hair available: http://puu.sh/iRGHV/a4c6a95861.png

There is also a minor note with glitching yourself unwanted hairstyles: whenever the "covering" type setting is used, it leaves the hair to be the same, at least on the preview image. This may be abused to give Unathi freakishly long hair, or stuff like that.

jack-fractal commented:
Hey Skull! 

Thanks for the testing. Yeah. I'd messed up the tail icon part of the preview. It was getting the default tajaran icon and overlaying it over the results. 

Fixed now.

I've also fixed the hairstyle thing. You can't end up with screwball hair styles in the previews anymore.

And I fixed the brain issue. It wasn't loading a default properly if you switched your character from one species too another.

All done! 

Hmm... I should probably get an IPC whitelist sometime soon...

jack-fractal commented:
Ack. I've got something else to remove. Hold on. I'm just checking the hair thing.

OK. Checked the hair thing. It looks like most of those hairstyles are also available to Unathi on the development branch. Not sure what's going on there. I have to remove something that disables the species whitelists.

jack-fractal commented:
OK. Species white-list reactivated. All good.

skull132 commented:
Alright. Then we're good to merge. I'll do it later today, or tomorrow, depending on how passed out I am.
